File: work/src/MIL/Examples/General/MdigGrabSequence/C++/mdiggrabsequence.cpp
/********************************************************************************/
/*
 * File name: MDigGrabSequence.cpp
 *
 * Synopsis:  This example shows how to grab a sequence, archive it, and play 
 *            it back in real time from an AVI file. 
 *
 * NOTE:      This example assumes that the hard disk is sufficiently fast 
 *            to keep up with the grab. Also, removing the sequence display or 
 *            the text annotation while grabbing will reduce the CPU usage and
 *            might help if some frames are missed during acquisition. 
 *            If the disk or system are not fast enough, set GRAB_SCALE to 0.5, 
 *            FRAME_NUMBER_ANNOTATION to M_NO or SAVE_SEQUENCE_TO_DISK to M_NO.
 */
#include <mil.h>

/* Sequence file name.*/
#define SEQUENCE_FILE M_TEMP_DIR MIL_TEXT("MilSequence.avi")

/* Image acquisition scale. */
#define GRAB_SCALE   1.0

/* Quantization factor to use during the compression.
   Valid values are 1 to 99 (higher to lower quality). 
*/
#define COMPRESSION_Q_FACTOR   50

/* Annotation flag. Set to M_YES to draw the frame number in the saved image. */
#define FRAME_NUMBER_ANNOTATION  M_YES

/* Archive flag. Set to M_NO to disable AVI Import/Export to disk. */
#define SAVE_SEQUENCE_TO_DISK  M_YES

/* Maximum number of images for the multiple buffering grab. */
#define NB_GRAB_IMAGE_MAX  22

/* User's archive hook function prototype (called for every grabbed frame). */
MIL_INT MFTYPE ArchiveFunction(MIL_INT HookType, MIL_ID HookId, void* HookDataPtr);

/* User's archive function hook data structure. */
typedef struct
   {
   MIL_ID MilSystem;
   MIL_ID MilDisplay;
   MIL_ID MilImageDisp;
   MIL_ID MilCompressedImage;
   MIL_INT NbGrabbedFrames;
   MIL_INT NbArchivedFrames;
   MIL_INT SaveSequenceToDisk;
   } HookDataStruct;


/* Main function. */
/* -------------- */
int MosMain(void)
{    
   MIL_ID     MilApplication, MilRemoteApplication, MilSystem, MilDigitizer, MilDisplay, MilImageDisp;
   MIL_ID     MilGrabImages[NB_GRAB_IMAGE_MAX];
   MIL_ID     MilCompressedImage = M_NULL;
   MIL_INT    CompressAttribute=0;
   MIL_INT    NbFrames=0,   Selection=1,   LicenseModules=0,   n=0;
   MIL_INT    FrameCount=0, FrameMissed=0, NbFramesReplayed=0, Exit=0;
   MIL_DOUBLE FrameRate=0,  TimeWait=0,    TotalReplay=0;
   MIL_DOUBLE GrabScale=GRAB_SCALE;
   MIL_INT    SaveSequenceToDisk = SAVE_SEQUENCE_TO_DISK;
   HookDataStruct UserHookData;   

   /* Allocate defaults. */
   MappAllocDefault(M_DEFAULT, &MilApplication, &MilSystem, &MilDisplay,
                                              &MilDigitizer, M_NULL);

   /* Allocate an image and display it. */
   MbufAllocColor(MilSystem,
                  MdigInquire(MilDigitizer, M_SIZE_BAND, M_NULL),
                  (MIL_INT)(MdigInquire(MilDigitizer, M_SIZE_X, M_NULL)*GrabScale),
                  (MIL_INT)(MdigInquire(MilDigitizer, M_SIZE_Y, M_NULL)*GrabScale),
                  8L+M_UNSIGNED,
                  M_IMAGE+M_GRAB+M_DISP, &MilImageDisp);
   MbufClear(MilImageDisp, 0x0);
   MdispSelect(MilDisplay, MilImageDisp);

   /* Grab continuously on display at the specified scale. */
   MdigControl(MilDigitizer, M_GRAB_SCALE, GrabScale);
   MdigGrabContinuous(MilDigitizer, MilImageDisp);

   /* Print a message */
   MosPrintf(MIL_TEXT("\nSEQUENCE ACQUISITION:\n"));
   MosPrintf(MIL_TEXT("--------------------\n\n"));

   /* Inquire MIL licenses. */
   MsysInquire(MilSystem, M_OWNER_APPLICATION, &MilRemoteApplication);
   MappInquire(MilRemoteApplication, M_LICENSE_MODULES, &LicenseModules);

   /* If sequence is saved to disk, select between grabbing an 
      uncompressed, JPEG or JPEG2000 sequence. 
   */
   if (SaveSequenceToDisk && (LicenseModules & (M_LICENSE_JPEGSTD | M_LICENSE_JPEG2000)))
      {
      MosPrintf(MIL_TEXT("Choose the sequence format:\n"));
      MosPrintf(MIL_TEXT("1) Uncompressed images.\n") );
      if(LicenseModules & M_LICENSE_JPEGSTD)
         MosPrintf(MIL_TEXT("2) Compressed images with a lossy JPEG algorithm.\n"));
      if(LicenseModules & M_LICENSE_JPEG2000)
         MosPrintf(MIL_TEXT("3) Compressed images with a lossy JPEG 2000 algorithm.\n"));
      Selection = MosGetch();
      }
   else
      {
      MosPrintf(MIL_TEXT("Press <Enter> to record images.\n"));
      Selection = '1';
      MosGetch();
      }


   /* Set the buffer attribute. */
   switch(Selection)
      {
      case '1':
      case '\r':
         MosPrintf(MIL_TEXT("\nRecording uncompressed images...\n\n"));
         CompressAttribute = M_NULL;
         break;

      case '2':
         MosPrintf(MIL_TEXT("\nRecording JPEG images...\n\n"));
         CompressAttribute = M_COMPRESS + M_JPEG_LOSSY;
         break;

      case '3':
         MosPrintf(MIL_TEXT("\nRecording JPEG 2000 images...\n\n"));
         CompressAttribute = M_COMPRESS + M_JPEG2000_LOSSY;
         break;

      default:
         MosPrintf(MIL_TEXT("\nInvalid selection !.\n\nUsing uncompressed images.\n\n"));
         CompressAttribute = M_NULL;
         while (MosKbhit()) MosGetch();
         break;
      }

   /* Allocate compressed buffer if required. */
   if (CompressAttribute)
      {
      MbufAllocColor(MilSystem,
                     MdigInquire(MilDigitizer, M_SIZE_BAND, M_NULL),
                     (MIL_INT)(MdigInquire(MilDigitizer, M_SIZE_X, M_NULL)*GrabScale),
                     (MIL_INT)(MdigInquire(MilDigitizer, M_SIZE_Y, M_NULL)*GrabScale),
                     8L+M_UNSIGNED, M_IMAGE+CompressAttribute, &MilCompressedImage);
      MbufControl(MilCompressedImage, M_Q_FACTOR, COMPRESSION_Q_FACTOR);
      }

   /* Allocate the grab buffers to hold the sequence buffering. */
   MappControl(M_DEFAULT, M_ERROR, M_PRINT_DISABLE);
   for (NbFrames=0, n=0; n<NB_GRAB_IMAGE_MAX; n++)
      {
      MbufAllocColor(MilSystem,
                     MdigInquire(MilDigitizer, M_SIZE_BAND, M_NULL),
                     (MIL_INT)(MdigInquire(MilDigitizer, M_SIZE_X, M_NULL)*GrabScale),
                     (MIL_INT)(MdigInquire(MilDigitizer, M_SIZE_Y, M_NULL)*GrabScale),
                     8L+M_UNSIGNED, M_IMAGE+M_GRAB, &MilGrabImages[n]);
      if (MilGrabImages[n])
        {
         NbFrames++;
         MbufClear(MilGrabImages[n], 0xFF);
        }
      else
         break;
      }
   MappControl(M_DEFAULT, M_ERROR, M_PRINT_ENABLE);

   /* Free buffers to leave space for possible temporary buffers. */
   for (n=0; n<2 && NbFrames; n++)
      {
      NbFrames--;
      MbufFree(MilGrabImages[NbFrames]);
      }

   /* Halt continuous grab. */
   MdigHalt(MilDigitizer);

   /* Open the AVI file if required. */
   if (SaveSequenceToDisk)
      {
      MosPrintf(MIL_TEXT("Saving the sequence to an AVI file...\n\n"));
      MbufExportSequence(SEQUENCE_FILE, M_DEFAULT, M_NULL, M_NULL, M_DEFAULT, M_OPEN);
      }

   /* Initialize User's archiving function hook data structure. */
   UserHookData.MilSystem           = MilSystem;
   UserHookData.MilDisplay          = MilDisplay;
   UserHookData.MilImageDisp        = MilImageDisp;
   UserHookData.MilCompressedImage  = MilCompressedImage;
   UserHookData.SaveSequenceToDisk  = SaveSequenceToDisk;
   UserHookData.NbGrabbedFrames     = 0;
   UserHookData.NbArchivedFrames    = 0;

   /* Acquire the sequence. The processing hook function will
      be called for each image grabbed to archive and display it. 
      If sequence is not saved to disk, stop after NbFrames.
   */
   MdigProcess(MilDigitizer, MilGrabImages, NbFrames, 
               SaveSequenceToDisk ? M_START : M_SEQUENCE, 
               M_DEFAULT, ArchiveFunction, &UserHookData);

   /* Wait for a key press. */
   MosPrintf(MIL_TEXT("Press <Enter> to continue.\n\n"));
   MosGetch();

   /* Stop the sequence acquisition. */
   MdigProcess(MilDigitizer, MilGrabImages, NbFrames, M_STOP, 
                             M_DEFAULT, ArchiveFunction, &UserHookData);

   /* Read and print final statistics. */
   MdigInquire(MilDigitizer, M_PROCESS_FRAME_COUNT,  &FrameCount);
   MdigInquire(MilDigitizer, M_PROCESS_FRAME_RATE,   &FrameRate);
   MdigInquire(MilDigitizer, M_PROCESS_FRAME_MISSED, &FrameMissed);
   MosPrintf(MIL_TEXT("\n\n%ld frames archived (%ld missed), at %.1f frames/sec ")
             MIL_TEXT("(%.1f ms/frame).\n"),UserHookData.NbArchivedFrames, 
                                           FrameMissed, FrameRate, 1000.0/FrameRate);

   /* Sequence file closing if required. */
   if (SaveSequenceToDisk)
       MbufExportSequence(SEQUENCE_FILE, M_DEFAULT, M_NULL, M_NULL, FrameRate, M_CLOSE);

   /* Playback the sequence until a key is pressed. */
   if (UserHookData.NbArchivedFrames > 0)
      {
      MIL_INT KeyPressed = 0;
      do
         {
         /* If sequence must be loaded. */
         if (SaveSequenceToDisk)
            {
            /* Inquire information about the sequence. */
            MosPrintf(MIL_TEXT("\nPlaying back sequence from the AVI file...\n"));
            MosPrintf(MIL_TEXT("Press <Enter> to end.\n\n"));
            MbufDiskInquire(SEQUENCE_FILE, M_NUMBER_OF_IMAGES, &FrameCount);
            MbufDiskInquire(SEQUENCE_FILE, M_FRAME_RATE, &FrameRate);
            MbufDiskInquire(SEQUENCE_FILE, M_COMPRESSION_TYPE, &CompressAttribute);

            /* Open the sequence file. */
            MbufImportSequence(SEQUENCE_FILE, M_DEFAULT, M_NULL, 
               M_NULL, M_NULL, M_NULL, M_NULL, M_OPEN);
            }
      
         /* Copy the images to the screen respecting the sequence frame rate. */
         TotalReplay = 0.0;
         NbFramesReplayed = 0;
         for (n=0; n<FrameCount; n++)
            {
            /* Reset the time. */
            MappTimer(M_DEFAULT, M_TIMER_RESET, M_NULL);

            /* If image was saved to disk. */
            if (SaveSequenceToDisk)
               {
               /* Load image directly to the display. */
               MbufImportSequence(SEQUENCE_FILE, M_DEFAULT, M_LOAD, 
                  M_NULL, &MilImageDisp, n, 1, M_READ);
               NbFramesReplayed++;
               MosPrintf(MIL_TEXT("Frame #%d             \r"), NbFramesReplayed);
               }
            else 
               {
               /* Copy the grabbed image to the display. */
               MbufCopy(MilGrabImages[n], MilImageDisp);
               NbFramesReplayed++;
               MosPrintf(MIL_TEXT("Frame #%d             \r"), NbFramesReplayed);
               }

            /* Check for a pressed key to exit. */
            if (MosKbhit() && (n>=(NB_GRAB_IMAGE_MAX-1)))
               {
               MosGetch();
               break;
               }

            /* Wait to have a proper frame rate. */
            MappTimer(M_DEFAULT, M_TIMER_READ, &TimeWait);
            TotalReplay += TimeWait;
            TimeWait = (1/FrameRate) - TimeWait;
            MappTimer(M_DEFAULT, M_TIMER_WAIT, &TimeWait);
            TotalReplay += (TimeWait > 0) ? TimeWait: 0.0;
            }

         /* Close the sequence file. */
         if (SaveSequenceToDisk)
            MbufImportSequence(SEQUENCE_FILE, M_DEFAULT, M_NULL, 
               M_NULL, M_NULL, M_NULL, M_NULL, M_CLOSE);

         /* Print statistics. */
         MosPrintf(MIL_TEXT("\n\n%ld frames replayed, at a frame rate of %.1f frames/sec ")
            MIL_TEXT("(%.1f ms/frame).\n\n"), NbFramesReplayed, n/TotalReplay, 
            1000.0*TotalReplay/n);
         MosPrintf(MIL_TEXT("Press <Enter> to end (or any other key to playback again).\n"));
         KeyPressed = MosGetch();
         }
      while((KeyPressed != '\r') && (KeyPressed != '\n'));
      }

   /* Free all allocated buffers. */
   MbufFree(MilImageDisp);
   for (n=0; n<NbFrames; n++)
       MbufFree(MilGrabImages[n]);
   if (MilCompressedImage)
      MbufFree(MilCompressedImage);

   /* Free defaults. */
   MappFreeDefault(MilApplication, MilSystem, MilDisplay, MilDigitizer, M_NULL);

   return 0;
}


/* User's archive function called each time a new buffer is grabbed. */
/* -------------------------------------------------------------------*/

/* Local defines for the annotations. */
#define STRING_LENGTH_MAX  20
#define STRING_POS_X       20
#define STRING_POS_Y       20

MIL_INT MFTYPE ArchiveFunction(MIL_INT HookType, MIL_ID HookId, void* HookDataPtr)
   {
   HookDataStruct *UserHookDataPtr = (HookDataStruct *)HookDataPtr;
   MIL_ID ModifiedImage = 0;
   MIL_INT n = 0;
   MIL_TEXT_CHAR Text[STRING_LENGTH_MAX]= {MIL_TEXT('\0'),};

   /* Retrieve the MIL_ID of the grabbed buffer. */
   MdigGetHookInfo(HookId, M_MODIFIED_BUFFER+M_BUFFER_ID, &ModifiedImage);

   /* Increment the frame count. */
   UserHookDataPtr->NbGrabbedFrames++;

   /* Draw the frame count in the image if enabled. */
   if ((FRAME_NUMBER_ANNOTATION == M_YES))
     {
     MosSprintf(Text, STRING_LENGTH_MAX, MIL_TEXT(" %ld "), 
                                          UserHookDataPtr->NbGrabbedFrames);
     MgraText(M_DEFAULT, ModifiedImage,   STRING_POS_X, STRING_POS_Y, Text);
     }

   /* Compress the new image. */
   if (UserHookDataPtr->MilCompressedImage)
      {
      MbufCopy(ModifiedImage, UserHookDataPtr->MilCompressedImage);
      }

   /* Archive the new image. */
   if (UserHookDataPtr->SaveSequenceToDisk)
      {
      MbufExportSequence(SEQUENCE_FILE, M_DEFAULT, UserHookDataPtr->MilCompressedImage? 
                         &UserHookDataPtr->MilCompressedImage: &ModifiedImage,
                         1, M_DEFAULT, M_WRITE);

      UserHookDataPtr->NbArchivedFrames++;
      MosPrintf(MIL_TEXT("Frame #%d               \r"), UserHookDataPtr->NbArchivedFrames);
      }

   /* Copy the new grabbed image to the display. */
   MbufCopy(ModifiedImage, UserHookDataPtr->MilImageDisp);

   return 0;
   }

File: work/src/MIL/Examples/General/MdigGrabSequence/VB/MdigGrabSequence.vb
'*******************************************************************************
'
' File name: MdigGrabSequence.vb
'
' Synopsis:  This example shows how to grab a sequence, archive it, and play 
'            it back in real time from an AVI file.
'
' NOTE:      This example assumes that the hard disk is sufficiently fast 
'            to keep up with the grab. Also, removing the sequence display or 
'            the text annotation while grabbing will reduce the CPU usage and
'            might help if some frames are missed during acquisition. 
'            If the disk or system are not fast enough, set GRAB_SCALE to 0.5, 
'            FRAME_NUMBER_ANNOTATION to False or SAVE_SEQUENCE_TO_DISK to False.
'*******************************************************************************
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports System.Runtime.InteropServices

Imports Matrox.MatroxImagingLibrary

Namespace MDigGrabSequence
    Friend Class Program
        ' Sequence file name
        Private Const SEQUENCE_FILE As String = MIL.M_TEMP_DIR & "MilSequence.avi"

        ' Image acquisition scale.
        Private Const GRAB_SCALE As Double = 1.0

        ' Quantization factor to use during the compression.
        ' Valid values are 1 to 99 (higher to lower quality).
        Private Const COMPRESSION_Q_FACTOR As Integer = 50

        ' Annotation flag. Set to False to draw the frame number in the saved image.
        Private Shared ReadOnly FRAME_NUMBER_ANNOTATION As MIL_INT = MIL.M_YES

        ' Archive flag. Set to False to disable AVI Import/Export to disk.
        Private Shared ReadOnly SaveSequenceToDisk As Boolean = True

        ' Maximum number of images for the multiple buffering grab.
        Private Const NB_GRAB_IMAGE_MAX As Integer = 22

        Public Class HookDataObject ' User's archive function hook data structure.
            Public MilSystem As MIL_ID
            Public MilDisplay As MIL_ID
            Public MilImageDisp As MIL_ID
            Public MilCompressedImage As MIL_ID
            Public NbGrabbedFrames As Integer
            Public NbArchivedFrames As Integer
            Public SaveSequenceToDisk As Boolean
        End Class

        ' Main function.
        ' --------------
        Shared Sub Main(ByVal args() As String)
            Dim MilApplication As MIL_ID = MIL.M_NULL
            Dim MilRemoteApplication As MIL_ID = MIL.M_NULL
            Dim MilSystem As MIL_ID = MIL.M_NULL
            Dim MilDigitizer As MIL_ID = MIL.M_NULL
            Dim MilDisplay As MIL_ID = MIL.M_NULL
            Dim MilImageDisp As MIL_ID = MIL.M_NULL
            Dim MilGrabImages(NB_GRAB_IMAGE_MAX) As MIL_ID
            Dim MilCompressedImage As MIL_ID = MIL.M_NULL
            Dim Selection As New ConsoleKeyInfo("1"c, ConsoleKey.D1, False, False, False)

            Dim NbFrames As Integer = 0
            Dim n As Integer = 0
            Dim NbFramesReplayed As Integer = 0
            Dim FrameRate As Double = 0
            Dim TimeWait As Double = 0
            Dim TotalReplay As Double = 0
            Dim GrabScale As Double = GRAB_SCALE
            Dim UserHookData As New HookDataObject()
            Dim LicenseModules As MIL_INT = 0
            Dim FrameCount As MIL_INT = 0
            Dim FrameMissed As MIL_INT = 0
            Dim CompressAttribute As MIL_INT = 0

            ' Allocate defaults.
            MIL.MappAllocDefault(MIL.M_DEFAULT, MilApplication, MilSystem, MilDisplay, MilDigitizer, CType(MIL.M_NULL, IntPtr))

            ' Allocate an image and display it.
            MIL.MbufAllocColor(MilSystem, MIL.MdigInquire(MilDigitizer, MIL.M_SIZE_BAND, MIL.M_NULL), CInt(CDbl(MIL.MdigInquire(MilDigitizer, MIL.M_SIZE_X, CType(MIL.M_NULL, IntPtr))) * GrabScale), CInt(CDbl(MIL.MdigInquire(MilDigitizer, MIL.M_SIZE_Y, CType(MIL.M_NULL, IntPtr))) * GrabScale), 8 + MIL.M_UNSIGNED, MIL.M_IMAGE + MIL.M_GRAB + MIL.M_DISP, MilImageDisp)

            MIL.MbufClear(MilImageDisp, &H0)
            MIL.MdispSelect(MilDisplay, MilImageDisp)

            ' Grab continuously on display at the specified scale.
            MIL.MdigControl(MilDigitizer, MIL.M_GRAB_SCALE, GrabScale)
            MIL.MdigGrabContinuous(MilDigitizer, MilImageDisp)

            ' Print a message
            Console.WriteLine()
            Console.WriteLine("SEQUENCE ACQUISITION:")
            Console.WriteLine("--------------------")
            Console.WriteLine()

            ' Inquire MIL licenses.
            MIL.MsysInquire(MilSystem, MIL.M_OWNER_APPLICATION, MilRemoteApplication)
            MIL.MappInquire(MilRemoteApplication, MIL.M_LICENSE_MODULES, LicenseModules)

            ' If sequence is saved to disk, select between grabbing an 
            ' uncompressed, JPEG or JPEG2000 sequence. 
            If SaveSequenceToDisk AndAlso ((LicenseModules And (MIL.M_LICENSE_JPEGSTD Or MIL.M_LICENSE_JPEG2000)) <> 0) Then
                Console.WriteLine("Choose the sequence format:")
                Console.WriteLine("1) Uncompressed images.")
                If (LicenseModules And MIL.M_LICENSE_JPEGSTD) <> 0 Then
                    Console.WriteLine("2) Compressed images with a lossy JPEG algorithm.")
                End If
                If (LicenseModules And MIL.M_LICENSE_JPEG2000) <> 0 Then
                    Console.WriteLine("3) Compressed images with a lossy JPEG 2000 algorithm.")
                End If
                Selection = Console.ReadKey()
            Else
                Console.WriteLine("Press <Enter> to record images.")
                Console.ReadKey()
            End If


            ' Set the buffer attribute.
            Select Case Selection.Key
                Case ConsoleKey.NumPad1, ConsoleKey.D1, ConsoleKey.Enter
                    Console.WriteLine()
                    Console.WriteLine("Recording uncompressed images...")
                    Console.WriteLine()
                    CompressAttribute = 0

                Case ConsoleKey.NumPad2, ConsoleKey.D2
                    Console.WriteLine()
                    Console.WriteLine("Recording JPEG images...")
                    Console.WriteLine()
                    CompressAttribute = MIL.M_COMPRESS + MIL.M_JPEG_LOSSY

                Case ConsoleKey.NumPad3, ConsoleKey.D3
                    Console.WriteLine()
                    Console.WriteLine("Recording JPEG 2000 images...")
                    Console.WriteLine()
                    CompressAttribute = MIL.M_COMPRESS + MIL.M_JPEG2000_LOSSY

                Case Else
                    Console.WriteLine()
                    Console.WriteLine("Invalid selection !.")
                    Console.WriteLine()
                    Console.WriteLine("Using uncompressed images.")
                    Console.WriteLine()
                    CompressAttribute = 0
                    Do While Console.KeyAvailable
                        Console.ReadKey()
                    Loop
            End Select

            ' Allocate a compressed buffer if required.
            If CompressAttribute <> MIL.M_NULL Then
                MIL.MbufAllocColor(MilSystem, MIL.MdigInquire(MilDigitizer, MIL.M_SIZE_BAND, MIL.M_NULL), CInt(CDbl(MIL.MdigInquire(MilDigitizer, MIL.M_SIZE_X, MIL.M_NULL)) * GrabScale), CInt(CDbl(MIL.MdigInquire(MilDigitizer, MIL.M_SIZE_Y, MIL.M_NULL)) * GrabScale), 8 + MIL.M_UNSIGNED, MIL.M_IMAGE + CompressAttribute, MilCompressedImage)

                MIL.MbufControl(MilCompressedImage, MIL.M_Q_FACTOR, COMPRESSION_Q_FACTOR)
            End If

            ' Allocate the grab buffers to hold the sequence buffering.
            MIL.MappControl(MIL.M_ERROR, MIL.M_PRINT_DISABLE)
            NbFrames = 0
            n = 0
            Do While n < NB_GRAB_IMAGE_MAX
                MIL.MbufAllocColor(MilSystem, MIL.MdigInquire(MilDigitizer, MIL.M_SIZE_BAND, MIL.M_NULL), CInt(CDbl(MIL.MdigInquire(MilDigitizer, MIL.M_SIZE_X, MIL.M_NULL)) * GrabScale), CInt(CDbl(MIL.MdigInquire(MilDigitizer, MIL.M_SIZE_Y, MIL.M_NULL)) * GrabScale), 8 + MIL.M_UNSIGNED, MIL.M_IMAGE + MIL.M_GRAB, MilGrabImages(n))

                If MilGrabImages(n) <> MIL.M_NULL Then
                    NbFrames += 1
                    MIL.MbufClear(MilGrabImages(n), &HFF)
                Else
                    Exit Do
                End If
                n += 1
            Loop
            MIL.MappControl(MIL.M_ERROR, MIL.M_PRINT_ENABLE)

            '  Free buffers to leave some space for possible temporary buffers.
            n = 0
            While n < 2 And NbFrames > 0
                NbFrames -= 1
                MIL.MbufFree(MilGrabImages(NbFrames))
                n += 1
            End While

            ' Halt continuous grab.
            MIL.MdigHalt(MilDigitizer)

            ' Open the AVI file if required.
            If SaveSequenceToDisk Then
                Console.WriteLine("Saving the sequence to an AVI file...")
                Console.WriteLine()
                MIL.MbufExportSequence(SEQUENCE_FILE, MIL.M_DEFAULT, CType(MIL.M_NULL, IntPtr), MIL.M_NULL, MIL.M_DEFAULT, MIL.M_OPEN)
            End If

            ' Initialize User's archiving function hook data structure.
            UserHookData.MilSystem = MilSystem
            UserHookData.MilDisplay = MilDisplay
            UserHookData.MilImageDisp = MilImageDisp
            UserHookData.MilCompressedImage = MilCompressedImage
            UserHookData.SaveSequenceToDisk = SaveSequenceToDisk
            UserHookData.NbGrabbedFrames = 0
            UserHookData.NbArchivedFrames = 0

            ' get a handle to the DigHookUserData object in the managed heap, we will use this 
            ' handle to get the object back in the callback function
            Dim UserHookDataHandle As GCHandle = GCHandle.Alloc(UserHookData)
            Dim UserHookFunctionDelegate As New MIL_DIG_HOOK_FUNCTION_PTR(AddressOf ArchiveFunction)

            ' Acquire the sequence. The processing hook function will
            ' be called for each image grabbed to archive and display it. 
            ' If sequence is not saved to disk, stop after NbFrames.
            If SaveSequenceToDisk Then
                MIL.MdigProcess(MilDigitizer, MilGrabImages, NbFrames, MIL.M_START, MIL.M_DEFAULT, UserHookFunctionDelegate, CType(UserHookDataHandle, IntPtr))
            Else
                MIL.MdigProcess(MilDigitizer, MilGrabImages, NbFrames, MIL.M_SEQUENCE, MIL.M_DEFAULT, UserHookFunctionDelegate, CType(UserHookDataHandle, IntPtr))
            End If

            ' Wait for a key press.
            Console.WriteLine("Press <Enter> to continue.")
            Console.WriteLine()
            Console.ReadKey(True)

            ' Stop the sequence acquisition.
            MIL.MdigProcess(MilDigitizer, MilGrabImages, NbFrames, MIL.M_STOP, MIL.M_DEFAULT, UserHookFunctionDelegate, CType(UserHookDataHandle, IntPtr))

            ' Free the GCHandle when no longer used.
            UserHookDataHandle.Free()

            ' Read and print final statistics.
            MIL.MdigInquire(MilDigitizer, MIL.M_PROCESS_FRAME_COUNT, FrameCount)
            MIL.MdigInquire(MilDigitizer, MIL.M_PROCESS_FRAME_RATE, FrameRate)
            MIL.MdigInquire(MilDigitizer, MIL.M_PROCESS_FRAME_MISSED, FrameMissed)

            Console.WriteLine()
            Console.WriteLine()
            Console.WriteLine("{0} frames archived ({1} missed), at {2:0.0} frames/sec ({3:0.0}ms/frame).", UserHookData.NbArchivedFrames, FrameMissed, FrameRate, 1000.0 / FrameRate)

            ' Sequence file closing if required.
            If SaveSequenceToDisk Then
                MIL.MbufExportSequence(SEQUENCE_FILE, MIL.M_DEFAULT, CType(MIL.M_NULL, IntPtr), MIL.M_NULL, FrameRate, MIL.M_CLOSE)
            End If

            ' Playback the sequence until a key is pressed.
            If UserHookData.NbArchivedFrames > 0 Then
                Do
                    ' If sequence must be loaded.
                    If SaveSequenceToDisk Then
                        ' Inquire information about the sequence.
                        Console.WriteLine()
                        Console.WriteLine("Playing back sequence from the AVI file...")
                        Console.WriteLine()
                        Console.WriteLine("Press <Enter> to end.")
                        Console.WriteLine()
                        Console.WriteLine()
                        MIL.MbufDiskInquire(SEQUENCE_FILE, MIL.M_NUMBER_OF_IMAGES, FrameCount)
                        MIL.MbufDiskInquire(SEQUENCE_FILE, MIL.M_FRAME_RATE, FrameRate)
                        MIL.MbufDiskInquire(SEQUENCE_FILE, MIL.M_COMPRESSION_TYPE, CompressAttribute)

                        ' Open the sequence file.
                        MIL.MbufImportSequence(SEQUENCE_FILE, MIL.M_DEFAULT, MIL.M_NULL, MIL.M_NULL, CType(MIL.M_NULL, IntPtr), MIL.M_NULL, MIL.M_NULL, MIL.M_OPEN)
                    End If

                    ' Copy the images to the screen respecting the sequence frame rate.
                    TotalReplay = 0.0
                    NbFramesReplayed = 0
                    n = 0
                    Do While n < FrameCount
                        ' Reset the time.
                        MIL.MappTimer(MIL.M_DEFAULT, MIL.M_TIMER_RESET, CType(MIL.M_NULL, IntPtr))

                        ' If image was saved to disk.
                        If SaveSequenceToDisk Then
                            ' Load image directly to the display.
                            MIL.MbufImportSequence(SEQUENCE_FILE, MIL.M_DEFAULT, MIL.M_LOAD, MIL.M_NULL, MilImageDisp, n, 1, MIL.M_READ)
                            NbFramesReplayed += 1
                            Console.Write("Frame #{0}             " & Constants.vbCr, NbFramesReplayed)
                        Else
                            ' Copy the grabbed image to the display.
                            MIL.MbufCopy(MilGrabImages(n), MilImageDisp)
                            NbFramesReplayed += 1
                            Console.Write("Frame #{0}             " & Constants.vbCr, NbFramesReplayed)

                        End If

                        ' Check for a pressed key to exit.
                        If Console.KeyAvailable AndAlso (n >= (NB_GRAB_IMAGE_MAX - 1)) Then
                            Console.ReadKey(True)
                            Exit Do
                        End If

                        ' Wait to have a proper frame rate.
                        MIL.MappTimer(MIL.M_DEFAULT, MIL.M_TIMER_READ, TimeWait)
                        TotalReplay += TimeWait
                        TimeWait = (1 / FrameRate) - TimeWait
                        MIL.MappTimer(MIL.M_DEFAULT, MIL.M_TIMER_WAIT, TimeWait)
                        If (TimeWait > 0) Then
                            TotalReplay += TimeWait
                        Else
                            TotalReplay += 0.0
                        End If
                        n += 1
                    Loop

                    ' Close the sequence file.
                    If SaveSequenceToDisk Then
                        MIL.MbufImportSequence(SEQUENCE_FILE, MIL.M_DEFAULT, MIL.M_NULL, MIL.M_NULL, CType(MIL.M_NULL, IntPtr), MIL.M_NULL, MIL.M_NULL, MIL.M_CLOSE)
                    End If

                    ' Print statistics.
                    Console.WriteLine()
                    Console.WriteLine()
                    Console.WriteLine("{0} frames replayed, at a frame rate of {1:0.0} frames/sec ({2:0.0} ms/frame).", NbFramesReplayed, n / TotalReplay, 1000.0 * TotalReplay / n)
                    Console.WriteLine()
                    Console.WriteLine("Press <Enter> to end (or any other key to playback again).")
                    Console.WriteLine()

                Loop While Console.ReadKey(True).Key <> ConsoleKey.Enter
            End If

            ' Free all allocated buffers.
            MIL.MbufFree(MilImageDisp)
            For n = 0 To NbFrames - 1
                MIL.MbufFree(MilGrabImages(n))
            Next n

            If MilCompressedImage <> MIL.M_NULL Then
                MIL.MbufFree(MilCompressedImage)
            End If

            ' Free defaults.
            MIL.MappFreeDefault(MilApplication, MilSystem, MilDisplay, MilDigitizer, MIL.M_NULL)
        End Sub

        ' User's archive function called each time a new buffer is grabbed.
        ' -------------------------------------------------------------------*/

        ' Local defines for the annotations.
        Private Const STRING_LENGTH_MAX As Integer = 20
        Private Const STRING_POS_X As Integer = 20
        Private Const STRING_POS_Y As Integer = 20

        Private Shared Function ArchiveFunction(ByVal HookType As MIL_INT, ByVal HookId As MIL_ID, ByVal HookDataPtr As IntPtr) As MIL_INT

            Dim HookDataHandle As GCHandle = CType(HookDataPtr, GCHandle)
            Dim UserHookDataPtr As HookDataObject = CType(HookDataHandle.Target, HookDataObject)
            Dim ModifiedImage As MIL_ID = MIL.M_NULL

            ' Retrieve the MIL_ID of the grabbed buffer.
            MIL.MdigGetHookInfo(HookId, MIL.M_MODIFIED_BUFFER + MIL.M_BUFFER_ID, ModifiedImage)

            ' Increment the frame count.
            UserHookDataPtr.NbGrabbedFrames += 1

            ' Draw the frame count in the image if enabled.
            If FRAME_NUMBER_ANNOTATION = MIL.M_YES Then
                MIL.MgraText(MIL.M_DEFAULT, ModifiedImage, STRING_POS_X, STRING_POS_Y, UserHookDataPtr.NbGrabbedFrames.ToString())
            End If

            ' Compress the new image if required while archiving the previous one.
            If UserHookDataPtr.MilCompressedImage <> MIL.M_NULL Then
                MIL.MbufCopy(ModifiedImage, UserHookDataPtr.MilCompressedImage)
            End If

            ' Archive the new image.
            If UserHookDataPtr.SaveSequenceToDisk Then
                Dim ImageToWrite As MIL_ID
                If UserHookDataPtr.MilCompressedImage <> MIL.M_NULL Then
                    ImageToWrite = UserHookDataPtr.MilCompressedImage
                Else
                    ImageToWrite = ModifiedImage
                End If

                MIL.MbufExportSequence(SEQUENCE_FILE, MIL.M_DEFAULT, ImageToWrite, 1, MIL.M_DEFAULT, MIL.M_WRITE)

                UserHookDataPtr.NbArchivedFrames += 1
                Console.Write("Frame #{0}               " & Constants.vbCr, UserHookDataPtr.NbArchivedFrames)
            End If

            ' Copy the new grabbed image to the display.
            MIL.MbufCopy(ModifiedImage, UserHookDataPtr.MilImageDisp)

            Return MIL.M_NULL
        End Function
    End Class
End Namespace
